src/lifecycle.js

```diff
@@ -124,7 +129,14 @@ function triggerAttributeChanged(target, component, data) {
   var type;
   var name = data.name;
   var newValue = data.newValue;
-  var oldValue = data.oldValue;
+
+  // Read the old attribute value from cache if needed, otherwise use native oldValue
+  var cachedAttributes;
+  if (needsAttrCaching) {
+    cachedAttributes = hasOwn(target, '_cachedAttrs') && target._cachedAttrs;
```

Comment:
Nitpick: something that describes the problem for the data name might be better. Naming is hard, though. `cachedAttributeOldValues`, `cachedOldValues`, `attributeOldValues`. On it's own, I'm not really sure what `cachedAttributes` really means.